openapi.yaml

```diff
@@ -471,6 +591,173 @@ paths:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        exposed:
+          type: boolean
+        exposed-endpoints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/ExposedEndpoint'
+        life:
+          type: string
+        principal:
+          type: boolean
+        remote:
+          type: boolean
+        tag:
+          type: string
+      additionalProperties: false
+      required:
+        - tag
+        - principal
+        - exposed
+        - remote
+        - life
+    ApplicationSetCharm:
+      type: object
+      properties:
+        application:
+          type: string
+        channel:
+          type: string
+        charm-origin:
+          $ref: '#/components/schemas/CharmOrigin'
+        charm-url:
+          type: string
+        config-settings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        config-settings-yaml:
+          type: string
+        endpoint-bindings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        force:
+          type: boolean
+        force-base:
+          type: boolean
+        force-units:
+          type: boolean
+        generation:
+          type: string
+        resource-ids:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        storage-constraints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/StorageConstraints'
+      additionalProperties: false
+      required:
+        - application
+        - generation
+        - charm-url
+        - channel
+        - force
+        - force-units
+        - force-base
+    Base:
+      type: object
+      properties:
+        channel:
+          type: string
+        name:
+          type: string
+      additionalProperties: false
+      required:
+        - name
+        - channel
+    CharmRelation:
+      type: object
+      properties:
+        interface:
+          type: string
+        limit:
+          type: integer
+        name:
+          type: string
+        optional:
+          type: boolean
+        role:
+          type: string
+        scope:
+          type: string
+      additionalProperties: false
+      required:
+        - name
+        - role
+        - interface
+        - optional
+        - limit
+        - scope
+    CharmOrigin:
+      type: object
+      properties:
+        architecture:
+          type: string
+        base:
+          $ref: '#/components/schemas/Base'
+        branch:
+          type: string
+        hash:
+          type: string
+        id:
+          type: string
+        instance-key:
+          type: string
+        revision:
+          type: integer
+        risk:
+          type: string
+        source:
+          type: string
```

Comment:
Is this a URI or something else? I'm not sure what this field is?

Reply:
https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/core/charm/origin.go#L16C46-L16C46

Reply:
Oh I'm sorry the link was a bit ambiguous, it's just an enum of ["local", "charmhub"]. [This](https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/core/charm/origin.go#L30) would be a better spot to show.

Reply:
Great, let's get that modelled as an enum then :)

Reply:
Already done 👍